### ISSUE TITLE
Timestamp via API is not accurate

### ISSUE THREAD

**Issue by deathbybandaid** [omarroth/invidious#1129]
# tldr from IRC (and updated for this posting):

At the time of discovery, the below video was 1-2 hours old via web and RSS, but API showed it as roughly a day old.

## Video Page time
https://www.invidio.us/watch?v=mo6Cyj0lI1Q

`Shared April 17, 2020`

## Channel RSS timestamp
https://www.invidio.us/feed/channel/UCdBK94H6oZT2Q7l0-b0xmMg

`2020-04-17T22:05:04+00:00`

## API timestamp in epoch
https://invidio.us/api/v1/videos/mo6Cyj0lI1Q?pretty=1

`"published": 1587081600`

When converted to a readable timestamp we get

`Friday, April 17, 2020 00:00:00 AM` (GMT)

## last thoughts

It looks like the API timestamp is roughly 22 hours off of actual posting time.

After looking at some other videos, it looks like RSS is consistently correct, but the API epoch is just midnight of the day, making timestamps inconsistently off.

If RSS is right, API should be as well?

**Comment by cloudrac3r** [comment]
Wow, this is really weird.

Look at this video on YouTube: https://www.youtube.com/watch?v=4qnNr24n4aA

Right now, it only has 343 views. That's not a lot for this channel. That means it must be very recently uploaded.

Using the Invidious API, `published` is `1587340800`, which at the time of writing is about 16 hours before the current UTC timestamp, `1587398703`. The publishedText on Invidious mirrors this: `"16 hours ago"`.

However, [when I look at the video on my website CloudTube,](https://cadence.moe/cloudtube/channel/UC3g50Yr3GFPYrBj-n55BdqA) it says 29 minutes ago, which seems like a much more likely time. However CloudTube gets its data from the Invidious API. What gives?

I opened up CloudTube's code and it gets its dates from the YouTube channel RSS feed: `https://www.youtube.com/feeds/videos.xml?${channelType}=${channelID}`

So you're completely correct. The API timestamp is wrong (by about 16 hours?) and the timestamp from YouTube's RSS is correct.

Well, a time being off by several hours in programming is probably the fault of time zones. But UTC time offsets only go up to ±13 hours, not 16. So I have no theories left as to why this happens.

**Comment by Perflyst** [comment]
> The publishedText on Invidious mirrors this: "16 hours ago"

Could this be the upload timestamp?

**Comment by cloudrac3r** [comment]
You mean upload vs publish to the public? I don't think so — IIRC all videos on that channel are scheduled into the future, but I don't think any are hidden and then made public later.

I believe Invidious gets its `published` value from converting the `publishedText` that it retrieves from youtube dot com into a timestamp, since YouTube doesn't actually show the exact timestamp on the watch page.

**Comment by deathbybandaid** [comment]
the odd thing I'm seeing is the API timestamp seems pretty consistently midnight of that day

**Comment by cloudrac3r** [comment]
If the video timestamp is X days ago, then accuracy to the day is the best that Invidious can do.

2. Many channels schedule videos that have already been recorded and edited for a consistent time of day every day, usually at a whole hour mark.

**Comment by deathbybandaid** [comment]
that logic seems off,,, if the page says 2 hours ago, then it shouldn't be close to 24 hours off in the timestamp, which is what I found the other day, and posted in IRC:

(Note: my IRC timestamps are Eastern Time)

````
20:03:52 <deathbybandaid> Just noticed an odd published time issue
20:04:18 <deathbybandaid> The new ShortCircuit episode gives me 1587081600 
20:05:16 <deathbybandaid> Which seems to be a full day behind the time of actual video publishing displayed on the video page
20:06:27 <deathbybandaid> Tldr: video is about 1hour old as of now the 17th, and the timestamp via api says it's the 16th
20:11:54 <deathbybandaid> 
````
![image](https://user-images.githubusercontent.com/18010545/79776060-2826e680-8303-11ea-885b-d41f48e93a23.png)

````
20:12:01 <deathbybandaid> 
````
![image](https://user-images.githubusercontent.com/18010545/79776103-383ec600-8303-11ea-8e1a-ce19086af97c.png)

````
20:12:16 <deathbybandaid> RSS feed seems correct with time too
20:13:18 <deathbybandaid> I even doublechecked via http://unixtimeconverter.everydaydevtools.com/
20:16:10 <deathbybandaid> The time of day seems correct-ish, but not the date
20:22:25 <deathbybandaid> Just doublechecked with https://invidio.us/api/v1/videos/mo6Cyj0lI1Q?pretty=1
20:22:36 <deathbybandaid> Same published time
20:23:36 <deathbybandaid> Here's the kicker:  
````
![image](https://user-images.githubusercontent.com/18010545/79776135-44c31e80-8303-11ea-801d-43fd5c949f86.png)

````

20:24:04 <deathbybandaid> The publishedtext says it's from 1 day ago
````

**Comment by deathbybandaid** [comment]
I understand if the RSS feed pulls a different timestamp than what the API pulls from the page.

However, if something on the page says "2 hours ago", that does not equate to "1 day ago" by any means. `time.current - 2 != time.midnight`

**Comment by deathbybandaid** [comment]
Optimally, the API would have a timestamp that is accurate to withing a minute,, I'd settle for being remotely close to the time of day

**Comment by omarroth** [comment]
See #570.

As mentioned in the linked issue, the `published` and `publishedText` are pulled from `datePublished`, which is in `%Y-%m-%d` format.

As @cloudrac3r mentioned, the RSS feed provides accurate published times for the most recent 15 uploads from a channel, but the watch page does not provide that information, so it isn't really practical as a general solution unfortunately..

**Comment by deathbybandaid** [comment]
HOLD THE PHONE!

just found some other interesting-ness regarding this! The Channel API gets a timestamp that seems a great deal more precise?

## Video API

https://invidio.us/api/v1/videos/mo6Cyj0lI1Q?pretty=1

````
  "published": 1587081600,
  "publishedText": "4 days ago",
````

## Channel API

https://invidio.us/api/v1/channels/videos/UCdBK94H6oZT2Q7l0-b0xmMg?pretty=1

````
    "published": 1587218020,
    "publishedText": "3 days ago",
````

**Comment by deathbybandaid** [comment]
One step further:

Linus Tech Tips is a pretty old channel, so let's `sort_by=oldest` and look at the timestamp for the first video of the channel.

## channel
https://invidio.us/api/v1/channels/videos/UCXuqSBlHAE6Xw-yeJA0Tunw?pretty=1&sort_by=oldest

````
    "published": 1240322705,
    "publishedText": "11 years ago",
````

## video

https://invidio.us/api/v1/videos/zGJSFTaFJbQ?pretty=1

````
  "published": 1227571200,
  "publishedText": "11 years ago",
````
